### PR TITLE
give 15 char limit when inputing player name in highscores

### LINKED ISSUE #4167: The player name (at least for the High Score purpose) should not exceed 15 chars
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

The High Score does not support player name longer than 15 chars. Please see row no 8 and 9:

image

### Save file

To end this game, get the Black Pearl artifact. Then the player name input is shown (at least if your score allows to place you on your High Score list).

[No_Name_21_0005.zip](https://github.com/ihhub/fheroes2/files/7111264/No_Name_21_0005.zip)

### Additional info

The OG accepts max 15 chars here.

### PR DESCRIPTION
**Pull request by eos428**
fix #4167

### CODE REVIEW

File: src/fheroes2/dialog/dialog_selectcount.cpp [hunk 1]
```diff
@@ -213,7 +213,7 @@ bool Dialog::SelectCount( const std::string & header, u32 min, u32 max, u32 & cu
     return result == Dialog::OK;
 }
 
-bool Dialog::InputString( const std::string & header, std::string & res, const std::string & title )
+bool Dialog::InputString( const std::string & header, std::string & res, const std::string & title, const size_t charLimit )
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog_selectcount.cpp:216]: [unrelated to this change but] Could we please change lines 225-226 into:
```cpp
res.clear();
```
?
The same check is done in `clear` call actually.

File: src/fheroes2/dialog/dialog.h [hunk 2]
```diff
@@ -103,7 +103,7 @@ namespace Dialog
     int ResourceInfo( const std::string &, const std::string &, const Funds &, int buttons = Dialog::OK );
     int SelectSkillFromArena( void );
     bool SelectCount( const std::string &, u32 min, u32 max, u32 & res, int step = 1 );
-    bool InputString( const std::string &, std::string &, const std::string & title = std::string() );
+    bool InputString( const std::string & header, std::string & result, const std::string & title = std::string(), const size_t charLimit = 99 );
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog.h:106]: What if we don't want to have a limit? Shall we set this value to **0** by default and modify the condition
```cpp
if ( charLimit > res.size() || le.KeyValue() == KeySym::KEY_BACKSPACE )
```
into
```cpp
if ( charLimit == 0 || charLimit > res.size() || le.KeyValue() == KeySym::KEY_BACKSPACE )
```
?

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @eos428 , I left few comments here. Would you mind to take a look?

File: src/fheroes2/dialog/dialog.h (hunk 2 above)
Comment (eos428) [reply src/fheroes2/dialog/dialog.h:106]: Yeah sure, this should work too. Originally I wanted -1 but I forgot `size_t` is unsigned :D

**Comment by ihhub** [comment]
@eos428 , thank you so much for this fix!

**Pull request merged**